### DISCUSSION TITLE
Accessing Network-Mapped Devices from Packaged WinRT Application

### DISCUSSION THREAD

**Discussion by lukeblevins** [microsoft/WindowsAppSDK#83] [General]
In Files, we use the FindFirstFileExFromApp API to enumerate storage locations. When accessing a network drive this way with the broadFileSystemAccess capability, it works as expected. However, when the app is compiled for release to the Microsoft Store, none of the files inside the drive are shown. Is there a capability I'm missing, or should Project Reunion improve the filesystem apis and/or relevant capabilities to allow for access to mapped NAS devices when packaged?

**Comment by lukeblevins** [comment]
@jonwis Do you have input on this scenario? It seems like I'm missing a capability, but it displays my NAS drive files perfectly fine in debug mode.

**Reply by lukeblevins** [reply]
It's similar to the issue described here: https://stackoverflow.com/questions/33090669/how-to-access-network-share-from-raspberry-pi-running-iot-core-in-uwp-app

**Reply by jonwis** [reply]
If `broadFileSystemAccess` isn't working for that I think we need @smaillet-ms 's advice.  In the meantime, can you use ProcMon to see what the app is accessing and maybe something will leap out as a difference? I'm assuming you've _enabled_ the `bFSA` capability for this app in settings?  

Links of note:

* https://docs.microsoft.com/en-us/windows/uwp/files/file-access-permissions
* https://github.com/MicrosoftDocs/windows-uwp/issues/2348
* https://github.com/microsoft/AppModelSamples/tree/master/Samples/BfsaFileBrowser

**Reply by lukeblevins** [reply]
@jonwis The app includes a FullTrust component also, so it is granted the BFSA capability automatically and doesn't appear in the list on versions 2004+

**Reply by lukeblevins** [reply]
I'm able to access other brokered locations. Just not the network mapped locations.

**Reply by lukeblevins** [reply]
@jonwis Some more findings.

- Accessing the mapped locations through `FindFirstFileExFromApp()` with the broadFileSystemAccess, privateNetworkClientServer, and internetClient capabilities will return an invalid handle (-1), when other non-network locations work fine.

- Marshal.GetLastWin32Error() returns code 65 `ERROR_NETWORK_ACCESS_DENIED`

- I, and other users, notice this issue with other third-party UWP File explorer apps.

- Attempting to create a handle for a subdirectory with the following:
` var h = _rootFolder.CreateSafeFileHandle("Captures", FileMode.Open);`
results in this stack trace:
```
Exception thrown at 0x00007FFB5A323EAC (KernelBase.dll) in Files.exe: WinRT originate error - 0x80070005 : 'System.UnauthorizedAccessException: Access to the path 'Captures' is denied.
   at System.IO.WindowsRuntimeStorageExtensions.CreateSafeFileHandle(IStorageFolder rootDirectory, String relativePath, FileMode mode, FileAccess access, FileShare share, FileOptions options)
   at System.IO.WindowsRuntimeStorageExtensions.CreateSafeFileHandle(IStorageFolder rootDirectory, String relativePath, FileMode mode)
   at Files.Filesystem.ItemViewModel.EnumerateItemsFromStandardFolder(String path)
   at Files.Filesyste'.
Exception thrown at 0x00007FFB5A323EAC (KernelBase.dll) in Files.exe: WinRT originate error - 0x80070005 : 'Access to the path 'Captures' is denied.'.
```

**This is only an issue when packaged,** as deploying a release or debug build directly from Visual Studio works flawlessly.

**Reply by jonwis** [reply]
Thanks.  @smaillet-ms - suggestions?

**Reply by smaillet-ms** [reply]
broadFileSystemAccess capability applies **ONLY** to the local file system. Network access requires additional capabilities, depending on the location it can require enterprise authentication etc... (Basically the app needs to have permissions to use the credentials needed to access the remote resource.) The storage APIs defer access checks on a UNC path to to the networking stack and firewall and I'm not very familiar with what they do internally to determine access.

**Reply by jonwis** [reply]
Ok, then @keith-horton, you're next!

**Reply by smaillet-ms** [reply]
According to a quick read of the code the UNC access check is double layered, the firewall is queried for the user and package to detect if the location is controlled with the privateNetworkClientServer capability AND additionally the enterpriseAuthentication capability is required.

**Reply by lukeblevins** [reply]
@smaillet-ms Just added the enterpriseAuthentication capability to my app, repackaged it, and it still doesn't work...

**Reply by lukeblevins** [reply]
I also tried to get around the issue by catching the invalid handle, and adding the respective StorageFolder to the Future Access List, but it doesn't work this way either.

**Reply by lukeblevins** [reply]
I would just use Windows.Storage to enumerate the location, but the FromApp methods are far faster. @jonwis @smaillet-ms I wonder if it has something to do with the FromApp method I'm using? The docs for it state that it abides by the UWP security model, yet it might only be for the local filesystem.

**Comment by sunjw** [comment]
Hi, @lukeblevins 
Does this problem solved? I have the same issue with CreateFileFromApp function to open files on SMB share.
